src/Proto3/Wire/Reverse/Internal.hs

```diff
@@ -0,0 +1,717 @@
+--   to the full builder state, which is used relatively infrequently.
+--
+--   * 'Int#': the total number of non-metadata bytes originally available
+--   within the current buffer before we started consuming them,
+--   *plus* the number of bytes actually written to previous buffers.
+--   We subtract the current unused figure to get the total bytes written.
+--
+--   * 'Double#': suitably-aligned scratch space for serialization
+--   of 'Double' and 'Float' values.
+--
+-- Though this choice imposes a small memory overhead in every buffer,
+-- it reduces the time and space required to save and restore metadata
+-- around the many non-inline calls that typically occur while writing
+-- data into the buffer.
+data BuildRState = BuildRState
+  { currentBuffer :: {-# UNPACK #-}!(P.MutableByteArray RealWorld)
+      -- ^ Specifies the current buffer.  Note that through this field
+      -- every @'StablePtr' ('IORef' 'BuildRState')@ keeps one buffer
+      -- reachable until that stable pointer is explicitly destroyed.
+  , sealedBuffers :: BL.ByteString
+      -- ^ Holds the bytes written to previous buffers.  We arrange for
+      -- this field to be in normal form (not just weak head normal form).
+      -- But to avoid redundant evaluation we do not mark it strict.
+  }
+
+-- | Allocates fields backward from offset 0 relative to some hypothetical
+-- address, yielding the total size and alignment requirements, respectively,
+-- along with the monadic return value.  The total size includes any padding
+-- at the end that is needed to make it a multiple of the overall alignment.
+allocateFields :: State (Int, Int) a -> (a, Int, Int)
+allocateFields fields = (x, size, align)
+  where
+    (x, (off, align)) = runState fields (0, 1)
+    size = mod off align - off
+      -- Aligns the overall size @- off@ by prepending @mod off align@ padding
+      -- bytes.  Because @mod off align@ is in @[0, align)@ we are are neither
+      -- removing bytes nor adding more than we need.  And for some @k@ we have
+      --
+      -- > mod off align == off + k * align
+      --
+      -- and therefore we achieve the precise alignment desired:
+      --
+      -- > size = (off + k * align) - off == k * align
+
+-- | Within the monadic context established by 'allocateFields',
+-- allocates one suitably-aligned field and returns its offset.
+-- The argument is only a proxy for its type; we never evaluate it,
+-- and therefore you may pass 'undefined'.
+--
+-- WARNING: We assume that 'max' is the same as 'lcm' for any pair of
+-- alignment values, so that we can avoid using 'lcm', which does not
+-- evaluate at compile time.  Compile-time evaluation help our speed.
+allocatePrimitiveField :: Storable a => a -> State (Int, Int) Int
+allocatePrimitiveField proxy = state $ \(off0, align0) ->
+  let width1 = sizeOf proxy
+      align1 = alignment proxy
+      unaligned = off0 - width1
+      off2 = unaligned - mod unaligned align1
```

Comment:
Why is `off2` aligned to `align1` instead of `align2`?

Reply:
There are only 2 offsets, the starting offset and the ending offset.  But align1 is neither the starting nor ending alignment; it is the alignment of the field by itself.  I'll work on better names...

Reply:
a4cabced06d5802e9d67308e4d910107d25f7516 renames as follows:
```
off0 -> prevOff
align0 -> prevAlign

off2 -> nextOff
align2 -> nextAlign

width1 -> fieldWidth
align1 -> fieldAlign
```